index.js

```diff
@@ -1,194 +1,289 @@
+    this.words = words
+    this.names = names
+    this.passwords = passwords
+    this.wordsTree = trees.arrayToTree(this.words, true, 3)
+    this.namesTree = trees.arrayToTree(this.names, true, 3)
+    this.passwordsTree = trees.arrayToTree(this.passwords, true, 3)
+    this.minimumLength = 0
+    this.maximumLength = 0
+
     this.specialSymbolsUser = ',.!#@*'
     this.specialSymbolsSip = '-.*()%'
-    this.penetration = '.*'
-    this.checker = checker
-    this.mellt = mellt
-    this.minTimeToCrack = 0
-    this.checker.allowed_symbols = '_-!"?$%^&*()+={}[]:;@\'~#|<>,.?\\/ '
-    this.checker.min_length = 0
-    this.checker.max_length = 0
-    this.upper = 0
-  }
-
-  /**
-   * Set minimal password length
-   * @param {number} length
-   */
-  setMinLength(length) {
-    this.checker.min_length = length
-  }
-
-  /**
-   * Set maximal password length
-   * @param {number} length
-   */
-  setMaxLength(length) {
-    this.checker.max_length = length
-  }
-
-  /**
-   * Set allowed letters password can contain
-   * @param {string} letters
-   */
-  setAllowedLetters(upperLetters = '', lowerLetters = '') {
-    this.lowerLetters = lowerLetters
-    this.upperLetters = upperLetters
-    this.checker.allowed_letters = this.upperLetters + this.lowerLetters
-  }
-
-  /**
-   * Set allowed numbers password can contain
-   * @param {string} numbers
-   */
-  setAllowedNumbers(numbers) {
-    this.checker.allowed_numbers = numbers
+    this.specialSymbolsPenetration = '.*'
+    this.minimumTimeToCrack = 0
+    this.minimumNumberOfUpperLetters = 0
   }
 
-  /**
-   * Set allowed symbols password can contain
-   * @param {string} symbols
-   */
-  setAllowedSpecialSymbols(symbols) {
-    this.checker.allowed_symbols = symbols
+  addValidator(name, validator) {
+    this.validators[name] = {
+      validate: validator
+    }
   }
 
-  /**
-   * Should check that the password only has letters from allowed_letters in it
-   * @param  {boolean} active true|false
-   */
-  checkLetters(active) {
-    this.checker.checkLetters(active)
+  setMinimumLength(minimumLength) {
+    this.minimumLength = minimumLength
+    if (minimumLength === 0) {
+      delete this.validators.minimumLength
+    } else {
+      this.addValidator('minimalLength', this.validateMinimalLength)
+    }
   }
 
-  /**
-   * Should check that the password has letter in it
-   * @param  {boolean} active true|false
-   */
-  requireLetters(active) {
-    this.checker.requireLetters(active)
+  setMaximumLength(maximumLength) {
+    this.maximumLength = maximumLength
+    if (maximumLength === 0) {
+      delete this.valiadors.maximumLength
+    } else {
+      this.addValidator('maximalLength', this.validateMaximalLength)
+    }
   }
 
-  /**
-   * Should check that the password only has numbers from allowed_numbers in it
-   * @param  {boolean} active true|false
-   */
-  checkNumbers(active) {
-    this.checker.checkNumbers(active)
+  setAllowedUpperLetter(upperLetters) {
```

Comment:
tady by jsi měl asi explicitně testovat if (typeof upperLetters === 'string')  protože na to pak v tichostí spoléháš  volanim metody toUpperCase. Konstrukci 
```
this.allowedLetters = this.allowedLowerLetters + this.allowedUpperLetters
```
můžes zkrátit jako 

```
this.allowedLetters += this.allowedUpperLetters
```

Reply:
Opraveno